gensim/models/doc2vec.py

```diff
@@ -794,6 +794,23 @@ def train(self, documents=None, corpus_file=None, total_examples=None, total_wor
 
         """
         kwargs = {}
+
+        # Check if both documents and corpus_file are None
+        if corpus_file is None and documents is None:
+            raise TypeError("Either one of corpus_file or documents value must be provided.")
+
+        # Check if both documents and corpus_file are not None
+        if not ((corpus_file is None) ^ (documents is None)):
+            raise TypeError("Instead provide value to either of corpus_file or documents parameter but not both.")
+
+        # Check if corpus_file is string type
+        if documents is None and not isinstance(corpus_file, string_types):
```

Comment:
If `corpus_file` must be a valid path to a file, then why not explicitly check for that, instead of checking the type?

```suggestion
        if documents is None and not os.path.isfile(corpus_file):
```

Reply:
Wouldn't that throw an exception for incorrect (non-string) `corpus_file`? It's an interesting idea for a try-catch block though.

Reply:
We could wrap it in a _is_valid_corpus(corpus_file) function for malformed input.

Given that this input validation will be happening in mulitple places, it's probably not a bad idea.

Reply:
@piskvorky @mpenkov I've kept it simple. I used `os.path.isfile` as suggested. It returns `False` at all type of invalid inputs including non-string / float or whatever trash value a user would pass. 

Also, to do this for fasttext, I think it will be better:
1. To create a separate PR
2. Wait until the this PR is finalised so that I can follow the same code approach there are well.

Reply:
I'm -1 on creating separate PRs for this, for several reasons:

1. The changes are tiny. They span several lines of code.
2. The changes will be identical for each model (e.g. doc2vec, fasttext, etc)
3. PR overhead (review, merge, release, document) is not worth the effort.
4. It's salami-slicing.

For your second point, sure, I think it's a good idea to wait until people come to a consensus before expanding your approach to the other models.

Reply:
@mpenkov Could you please review the latest changes and let me know is we are good to go or something else needs to be done ?

Reply:
Sure. Left you some nitpicky comments.

In the future, please consider ticking "allow commits from maintainers" when creating a PR on github. That allows the reviewers of your PR to push changes to your branch, potentially shortcutting some of the back-and-forth for minor issues like typos, formatting, etc.

Reply:
@mpenkov thanks for the suggestion. I believe we are solid now on this PR. I will now make these changes in fasttext module as well.

Reply:
Yes, please go ahead.

Reply:
@mpenkov I've done the changes in fasttext module as well. Please have a look.